File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/GlobalPreferencesParser.java
/*
 * This file is part of BOINC.
 * http://boinc.berkeley.edu
 * Copyright (C) 2012 University of California
 *
 * BOINC is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 *
 * BOINC is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with BOINC.  If not, see <http://www.gnu.org/licenses/>.
 */
package edu.berkeley.boinc.rpc;

import org.xml.sax.Attributes;
import org.xml.sax.SAXException;

import android.util.Log;
import android.util.Xml;

import edu.berkeley.boinc.utils.Logging;

public class GlobalPreferencesParser extends BaseParser {
    static final String GLOBAL_PREFERENCES_TAG = "global_preferences";
    static final String DAY_PREFS_TAG = "day_prefs";
    static final String DAY_OF_WEEK_TAG = "day_of_week";
    static final String NET_START_HOUR_TAG = "net_start_hour";
    static final String NET_END_HOUR_TAG = "net_end_hour";

    private GlobalPreferences mPreferences = null;

    private boolean mInsideDayPrefs = false;
    private int mDayOfWeek = 0;
    private TimeSpan mTempCpuTimeSpan = null;
    private TimeSpan mTempNetTimeSpan = null;

    GlobalPreferences getGlobalPreferences() {
        return mPreferences;
    }

    public static GlobalPreferences parse(String rpcResult) {
        try {
            GlobalPreferencesParser parser = new GlobalPreferencesParser();
            Xml.parse(rpcResult, parser);
            return parser.getGlobalPreferences();
        }
        catch(SAXException e) {
            return null;
        }
    }

    @Override
    public void startElement(String uri, String localName, String qName, Attributes attributes) throws SAXException {
        super.startElement(uri, localName, qName, attributes);
        if(localName.equalsIgnoreCase(GLOBAL_PREFERENCES_TAG)) {
            mPreferences = new GlobalPreferences();
        }
        else if(localName.equalsIgnoreCase(DAY_PREFS_TAG)) {
            mInsideDayPrefs = true;
        }
        else {
            // Another element, hopefully primitive and not constructor
            // (although unknown constructor does not hurt, because there will be primitive start anyway)
            mElementStarted = true;
            mCurrentElement.setLength(0);
        }
    }

    @Override
    public void endElement(String uri, String localName, String qName) throws SAXException {
        super.endElement(uri, localName, qName);
        try {
            if(mPreferences != null) {
                // we are inside <global_preferences>
                // Closing tag of <global_preferences> - nothing to do at the moment
                if(!localName.equalsIgnoreCase(GLOBAL_PREFERENCES_TAG) &&
                   localName.equalsIgnoreCase(DAY_PREFS_TAG)) {
                    // closing <day_prefs>
                    if(mDayOfWeek >= 0 && mDayOfWeek <= 6) {
                        mPreferences.getCpuTimes().getWeekPrefs()[mDayOfWeek] = mTempCpuTimeSpan;
                        mPreferences.getNetTimes().getWeekPrefs()[mDayOfWeek] = mTempNetTimeSpan;
                    }

                    mTempCpuTimeSpan = null;
                    mTempNetTimeSpan = null;
                    mInsideDayPrefs = false;
                }
                else if(mInsideDayPrefs) {
                    trimEnd();
                    if(localName.equalsIgnoreCase(DAY_OF_WEEK_TAG)) {
                        mDayOfWeek = Integer.parseInt(mCurrentElement.toString());
                    }
                    else if(localName.equalsIgnoreCase(TimePreferences.Fields.START_HOUR)) {
                        if(mTempCpuTimeSpan == null) {
                            mTempCpuTimeSpan = new TimeSpan();
                        }
                        mTempCpuTimeSpan.setStartHour(Double.parseDouble(mCurrentElement.toString()));
                    }
                    else if(localName.equalsIgnoreCase(TimePreferences.Fields.END_HOUR)) {
                        if(mTempCpuTimeSpan == null) {
                            mTempCpuTimeSpan = new TimeSpan();
                        }
                        mTempCpuTimeSpan.setEndHour(Double.parseDouble(mCurrentElement.toString()));
                    }
                    else if(localName.equalsIgnoreCase(NET_START_HOUR_TAG)) {
                        if(mTempNetTimeSpan == null) {
                            mTempNetTimeSpan = new TimeSpan();
                        }
                        mTempNetTimeSpan.setStartHour(Double.parseDouble(mCurrentElement.toString()));
                    }
                    else if(localName.equalsIgnoreCase(NET_END_HOUR_TAG)) {
                        if(mTempNetTimeSpan == null) {
                            mTempNetTimeSpan = new TimeSpan();
                        }
                        mTempNetTimeSpan.setEndHour(Double.parseDouble(mCurrentElement.toString()));
                    }
                }
                else {
                    // Not the closing tag - we decode possible inner tags
                    trimEnd();
                    if(localName.equalsIgnoreCase(GlobalPreferences.Fields.RUN_ON_BATTERIES)) {
                        mPreferences.setRunOnBatteryPower(Integer.parseInt(mCurrentElement.toString()) != 0);
                    }
                    else if(localName.equalsIgnoreCase(GlobalPreferences.Fields.BATTERY_CHARGE_MIN_PCT)) {
                        mPreferences.setBatteryChargeMinPct(Double.parseDouble(mCurrentElement.toString()));
                    }
                    else if(localName.equalsIgnoreCase(GlobalPreferences.Fields.BATTERY_MAX_TEMPERATURE)) {
                        mPreferences.setBatteryMaxTemperature(Double.parseDouble(mCurrentElement.toString()));
                    }
                    else if(localName.equalsIgnoreCase(GlobalPreferences.Fields.RUN_GPU_IF_USER_ACTIVE)) {
                        mPreferences.setRunGpuIfUserActive(Integer.parseInt(mCurrentElement.toString()) != 0);
                    }
                    else if(localName.equalsIgnoreCase(GlobalPreferences.Fields.RUN_IF_USER_ACTIVE)) {
                        mPreferences.setRunIfUserActive(Integer.parseInt(mCurrentElement.toString()) != 0);
                    }
                    else if(localName.equalsIgnoreCase(GlobalPreferences.Fields.IDLE_TIME_TO_RUN)) {
                        mPreferences.setIdleTimeToRun(Double.parseDouble(mCurrentElement.toString()));
                    }
                    else if(localName.equalsIgnoreCase(GlobalPreferences.Fields.SUSPEND_CPU_USAGE)) {
                        mPreferences.setSuspendCpuUsage(Double.parseDouble(mCurrentElement.toString()));
                    }
                    else if(localName.equalsIgnoreCase(GlobalPreferences.Fields.LEAVE_APPS_IN_MEMORY)) {
                        mPreferences.setLeaveAppsInMemory(Integer.parseInt(mCurrentElement.toString()) != 0);
                    }
                    else if(localName.equalsIgnoreCase(GlobalPreferences.Fields.DONT_VERIFY_IMAGES)) {
                        mPreferences.setDoNotVerifyImages(Integer.parseInt(mCurrentElement.toString()) != 0);
                    }
                    else if(localName.equalsIgnoreCase(GlobalPreferences.Fields.WORK_BUF_MIN_DAYS)) {
                        mPreferences.setWorkBufMinDays(
                                Math.max(0.00001, Double.parseDouble(mCurrentElement.toString())));
                    }
                    else if(localName.equalsIgnoreCase(GlobalPreferences.Fields.WORK_BUF_ADDITIONAL_DAYS)) {
                        mPreferences.setWorkBufAdditionalDays(
                                Math.max(0.0, Double.parseDouble(mCurrentElement.toString())));
                    }
                    else if(localName.equalsIgnoreCase(GlobalPreferences.Fields.MAX_NCPUS_PCT)) {
                        mPreferences.setMaxNoOfCPUsPct(Double.parseDouble(mCurrentElement.toString()));
                    }
                    else if(localName.equalsIgnoreCase(GlobalPreferences.Fields.CPU_SCHEDULING_PERIOD_MINUTES)) {
                        double value = Double.parseDouble(mCurrentElement.toString());
                        if (value < 0.00001) {
                            value = 60;
                        }
                        mPreferences.setCpuSchedulingPeriodMinutes(value);
                    }
                    else if(localName.equalsIgnoreCase(GlobalPreferences.Fields.DISK_INTERVAL)) {
                        mPreferences.setDiskInterval(Double.parseDouble(mCurrentElement.toString()));
                    }
                    else if(localName.equalsIgnoreCase(GlobalPreferences.Fields.DISK_MAX_USED_GB)) {
                        mPreferences.setDiskMaxUsedGB(Double.parseDouble(mCurrentElement.toString()));
                    }
                    else if(localName.equalsIgnoreCase(GlobalPreferences.Fields.DISK_MAX_USED_PCT)) {
                        mPreferences.setDiskMaxUsedPct(Double.parseDouble(mCurrentElement.toString()));
                    }
                    else if(localName.equalsIgnoreCase(GlobalPreferences.Fields.DISK_MIN_FREE_GB)) {
                        mPreferences.setDiskMinFreeGB(Double.parseDouble(mCurrentElement.toString()));
                    }
                    else if(localName.equalsIgnoreCase(GlobalPreferences.Fields.RAM_MAX_USED_BUSY_FRAC)) {
                        mPreferences.setRamMaxUsedBusyFrac(Double.parseDouble(mCurrentElement.toString()));
                    }
                    else if(localName.equalsIgnoreCase(GlobalPreferences.Fields.RAM_MAX_USED_IDLE_FRAC)) {
                        mPreferences.setRamMaxUsedIdleFrac(Double.parseDouble(mCurrentElement.toString()));
                    }
                    else if(localName.equalsIgnoreCase(GlobalPreferences.Fields.MAX_BYTES_SEC_UP)) {
                        mPreferences.setMaxBytesSecUp(Double.parseDouble(mCurrentElement.toString()));
                    }
                    else if(localName.equalsIgnoreCase(GlobalPreferences.Fields.MAX_BYTES_SEC_DOWN)) {
                        mPreferences.setMaxBytesSecDown(Double.parseDouble(mCurrentElement.toString()));
                    }
                    else if(localName.equalsIgnoreCase(GlobalPreferences.Fields.CPU_USAGE_LIMIT)) {
                        mPreferences.setCpuUsageLimit(Double.parseDouble(mCurrentElement.toString()));
                    }
                    else if(localName.equalsIgnoreCase(GlobalPreferences.Fields.DAILY_XFER_PERIOD_MB)) {
                        mPreferences.setDailyTransferLimitMB(Double.parseDouble(mCurrentElement.toString()));
                    }
                    else if(localName.equalsIgnoreCase(GlobalPreferences.Fields.DAILY_XFER_PERIOD_DAYS)) {
                        mPreferences.setDailyTransferPeriodDays(Integer.parseInt(mCurrentElement.toString()));
                    }
                    else if(localName.equalsIgnoreCase(TimePreferences.Fields.START_HOUR)) {
                        mPreferences.getCpuTimes().setStartHour(Double.parseDouble(mCurrentElement.toString()));
                    }
                    else if(localName.equalsIgnoreCase(TimePreferences.Fields.END_HOUR)) {
                        mPreferences.getCpuTimes().setEndHour(Double.parseDouble(mCurrentElement.toString()));
                    }
                    else if(localName.equalsIgnoreCase(NET_START_HOUR_TAG)) {
                        mPreferences.getNetTimes().setStartHour(Double.parseDouble(mCurrentElement.toString()));
                    }
                    else if(localName.equalsIgnoreCase(NET_END_HOUR_TAG)) {
                        mPreferences.getNetTimes().setEndHour(Double.parseDouble(mCurrentElement.toString()));
                    }
                    else if(localName.equalsIgnoreCase(GlobalPreferences.Fields.OVERRIDE_FILE_PRESENT)) {
                        mPreferences.setOverrideFilePresent(Integer.parseInt(mCurrentElement.toString()) != 0);
                    }
                    else if(localName.equalsIgnoreCase(GlobalPreferences.Fields.NETWORK_WIFI_ONLY)) {
                        mPreferences.setNetworkWiFiOnly(Integer.parseInt(mCurrentElement.toString()) != 0);
                    }
                }
            }
        }
        catch(NumberFormatException e) {
            if(Logging.ERROR.equals(Boolean.TRUE)) {
                Log.e(Logging.TAG, "GlobalPreferencesParser.endElement error: ", e);
            }
        }
        mElementStarted = false;
    }
}

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/GlobalPreferencesParser.kt
/*
 * This file is part of BOINC.
 * http://boinc.berkeley.edu
 * Copyright (C) 2021 University of California
 *
 * BOINC is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 *
 * BOINC is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with BOINC.  If not, see <http://www.gnu.org/licenses/>.
 */
package edu.berkeley.boinc.rpc

import android.util.Xml
import edu.berkeley.boinc.utils.Logging
import org.xml.sax.Attributes
import org.xml.sax.SAXException

class GlobalPreferencesParser : BaseParser() {
    lateinit var globalPreferences: GlobalPreferences
        private set
    private var mInsideDayPrefs = false
    private var mDayOfWeek = 0
    private var mTempCpuTimeSpan: TimeSpan? = null
    private var mTempNetTimeSpan: TimeSpan? = null

    @Throws(SAXException::class)
    override fun startElement(uri: String?, localName: String, qName: String?, attributes: Attributes?) {
        super.startElement(uri, localName, qName, attributes)
        when {
            localName.equals(GLOBAL_PREFERENCES_TAG, ignoreCase = true) -> {
                globalPreferences = GlobalPreferences()
            }
            localName.equals(DAY_PREFS_TAG, ignoreCase = true) -> {
                mInsideDayPrefs = true
            }
            else -> { // Another element, hopefully primitive and not constructor
                // (although unknown constructor does not hurt, because there will be primitive start anyway)
                mElementStarted = true
                mCurrentElement.setLength(0)
            }
        }
    }

    @Throws(SAXException::class)
    override fun endElement(uri: String?, localName: String, qName: String?) {
        super.endElement(uri, localName, qName)
        try {
            if (!localName.equals(GLOBAL_PREFERENCES_TAG, ignoreCase = true) &&
                    localName.equals(DAY_PREFS_TAG, ignoreCase = true)) { // closing <day_prefs>
                if (mDayOfWeek in 0..6) {
                    globalPreferences.cpuTimes.weekPrefs[mDayOfWeek] = mTempCpuTimeSpan
                    globalPreferences.netTimes.weekPrefs[mDayOfWeek] = mTempNetTimeSpan
                }
                mTempCpuTimeSpan = null
                mTempNetTimeSpan = null
                mInsideDayPrefs = false
            } else if (mInsideDayPrefs) {
                trimEnd()
                when {
                    localName.equals(DAY_OF_WEEK_TAG, ignoreCase = true) -> {
                        mDayOfWeek = mCurrentElement.toInt()
                    }
                    localName.equals(TimePreferences.Fields.START_HOUR, ignoreCase = true) -> {
                        if (mTempCpuTimeSpan == null) {
                            mTempCpuTimeSpan = TimeSpan()
                        }
                        mTempCpuTimeSpan!!.startHour = mCurrentElement.toDouble()
                    }
                    localName.equals(TimePreferences.Fields.END_HOUR, ignoreCase = true) -> {
                        if (mTempCpuTimeSpan == null) {
                            mTempCpuTimeSpan = TimeSpan()
                        }
                        mTempCpuTimeSpan!!.endHour = mCurrentElement.toDouble()
                    }
                    localName.equals(NET_START_HOUR_TAG, ignoreCase = true) -> {
                        if (mTempNetTimeSpan == null) {
                            mTempNetTimeSpan = TimeSpan()
                        }
                        mTempNetTimeSpan!!.startHour = mCurrentElement.toDouble()
                    }
                    localName.equals(NET_END_HOUR_TAG, ignoreCase = true) -> {
                        if (mTempNetTimeSpan == null) {
                            mTempNetTimeSpan = TimeSpan()
                        }
                        mTempNetTimeSpan!!.endHour = mCurrentElement.toDouble()
                    }
                }
            } else { // Not the closing tag - we decode possible inner tags
                mCurrentElement.trimEnd()
                when {
                    localName.equals(GlobalPreferences.Fields.RUN_ON_BATTERIES, ignoreCase = true) -> {
                        globalPreferences.runOnBatteryPower = mCurrentElement.toInt() != 0
                    }
                    localName.equals(GlobalPreferences.Fields.BATTERY_CHARGE_MIN_PCT, ignoreCase = true) -> {
                        globalPreferences.batteryChargeMinPct = mCurrentElement.toDouble()
                    }
                    localName.equals(GlobalPreferences.Fields.BATTERY_MAX_TEMPERATURE, ignoreCase = true) -> {
                        globalPreferences.batteryMaxTemperature = mCurrentElement.toDouble()
                    }
                    localName.equals(GlobalPreferences.Fields.RUN_GPU_IF_USER_ACTIVE, ignoreCase = true) -> {
                        globalPreferences.runGpuIfUserActive = mCurrentElement.toInt() != 0
                    }
                    localName.equals(GlobalPreferences.Fields.RUN_IF_USER_ACTIVE, ignoreCase = true) -> {
                        globalPreferences.runIfUserActive = mCurrentElement.toInt() != 0
                    }
                    localName.equals(GlobalPreferences.Fields.IDLE_TIME_TO_RUN, ignoreCase = true) -> {
                        globalPreferences.idleTimeToRun = mCurrentElement.toDouble()
                    }
                    localName.equals(GlobalPreferences.Fields.SUSPEND_CPU_USAGE, ignoreCase = true) -> {
                        globalPreferences.suspendCpuUsage = mCurrentElement.toDouble()
                    }
                    localName.equals(GlobalPreferences.Fields.LEAVE_APPS_IN_MEMORY, ignoreCase = true) -> {
                        globalPreferences.leaveAppsInMemory = mCurrentElement.toInt() != 0
                    }
                    localName.equals(GlobalPreferences.Fields.DONT_VERIFY_IMAGES, ignoreCase = true) -> {
                        globalPreferences.doNotVerifyImages = mCurrentElement.toInt() != 0
                    }
                    localName.equals(GlobalPreferences.Fields.WORK_BUF_MIN_DAYS, ignoreCase = true) -> {
                        globalPreferences.workBufMinDays = 0.00001.coerceAtLeast(mCurrentElement.toDouble())
                    }
                    localName.equals(GlobalPreferences.Fields.WORK_BUF_ADDITIONAL_DAYS, ignoreCase = true) -> {
                        globalPreferences.workBufAdditionalDays = 0.0.coerceAtLeast(mCurrentElement.toDouble())
                    }
                    localName.equals(GlobalPreferences.Fields.MAX_NCPUS_PCT, ignoreCase = true) -> {
                        globalPreferences.maxNoOfCPUsPct = mCurrentElement.toDouble()
                    }
                    localName.equals(GlobalPreferences.Fields.CPU_SCHEDULING_PERIOD_MINUTES, ignoreCase = true) -> {
                        var value = mCurrentElement.toDouble()
                        if (value < 0.00001) {
                            value = 60.0
                        }
                        globalPreferences.cpuSchedulingPeriodMinutes = value
                    }
                    localName.equals(GlobalPreferences.Fields.DISK_INTERVAL, ignoreCase = true) -> {
                        globalPreferences.diskInterval = mCurrentElement.toDouble()
                    }
                    localName.equals(GlobalPreferences.Fields.DISK_MAX_USED_GB, ignoreCase = true) -> {
                        globalPreferences.diskMaxUsedGB = mCurrentElement.toDouble()
                    }
                    localName.equals(GlobalPreferences.Fields.DISK_MAX_USED_PCT, ignoreCase = true) -> {
                        globalPreferences.diskMaxUsedPct = mCurrentElement.toDouble()
                    }
                    localName.equals(GlobalPreferences.Fields.DISK_MIN_FREE_GB, ignoreCase = true) -> {
                        globalPreferences.diskMinFreeGB = mCurrentElement.toDouble()
                    }
                    localName.equals(GlobalPreferences.Fields.RAM_MAX_USED_BUSY_FRAC, ignoreCase = true) -> {
                        globalPreferences.ramMaxUsedBusyFrac = mCurrentElement.toDouble()
                    }
                    localName.equals(GlobalPreferences.Fields.RAM_MAX_USED_IDLE_FRAC, ignoreCase = true) -> {
                        globalPreferences.ramMaxUsedIdleFrac = mCurrentElement.toDouble()
                    }
                    localName.equals(GlobalPreferences.Fields.MAX_BYTES_SEC_UP, ignoreCase = true) -> {
                        globalPreferences.maxBytesSecUp = mCurrentElement.toDouble()
                    }
                    localName.equals(GlobalPreferences.Fields.MAX_BYTES_SEC_DOWN, ignoreCase = true) -> {
                        globalPreferences.maxBytesSecDown = mCurrentElement.toDouble()
                    }
                    localName.equals(GlobalPreferences.Fields.CPU_USAGE_LIMIT, ignoreCase = true) -> {
                        globalPreferences.cpuUsageLimit = mCurrentElement.toDouble()
                    }
                    localName.equals(GlobalPreferences.Fields.DAILY_XFER_PERIOD_MB, ignoreCase = true) -> {
                        globalPreferences.dailyTransferLimitMB = mCurrentElement.toDouble()
                    }
                    localName.equals(GlobalPreferences.Fields.DAILY_XFER_PERIOD_DAYS, ignoreCase = true) -> {
                        globalPreferences.dailyTransferPeriodDays = mCurrentElement.toInt()
                    }
                    localName.equals(TimePreferences.Fields.START_HOUR, ignoreCase = true) -> {
                        globalPreferences.cpuTimes.startHour = mCurrentElement.toDouble()
                    }
                    localName.equals(TimePreferences.Fields.END_HOUR, ignoreCase = true) -> {
                        globalPreferences.cpuTimes.endHour = mCurrentElement.toDouble()
                    }
                    localName.equals(NET_START_HOUR_TAG, ignoreCase = true) -> {
                        globalPreferences.netTimes.startHour = mCurrentElement.toDouble()
                    }
                    localName.equals(NET_END_HOUR_TAG, ignoreCase = true) -> {
                        globalPreferences.netTimes.endHour = mCurrentElement.toDouble()
                    }
                    localName.equals(GlobalPreferences.Fields.OVERRIDE_FILE_PRESENT, ignoreCase = true) -> {
                        globalPreferences.overrideFilePresent = mCurrentElement.toInt() != 0
                    }
                    localName.equals(GlobalPreferences.Fields.NETWORK_WIFI_ONLY, ignoreCase = true) -> {
                        globalPreferences.networkWiFiOnly = mCurrentElement.toInt() != 0
                    }
                }
            }
        } catch (e: Exception) {
            Logging.logException(Logging.Category.XML, "GlobalPreferencesParser.endElement error: ", e)
        }
        mElementStarted = false
    }

    companion object {
        const val GLOBAL_PREFERENCES_TAG = "global_preferences"
        const val DAY_PREFS_TAG = "day_prefs"
        const val DAY_OF_WEEK_TAG = "day_of_week"
        const val NET_START_HOUR_TAG = "net_start_hour"
        const val NET_END_HOUR_TAG = "net_end_hour"
        @JvmStatic
        fun parse(rpcResult: String?): GlobalPreferences? {
            return try {
                val parser = GlobalPreferencesParser()
                Xml.parse(rpcResult, parser)
                parser.globalPreferences
            } catch (e: SAXException) {
                Logging.logException(Logging.Category.RPC, "GlobalPreferencesParser: malformed XML ", e)
                Logging.logDebug(Logging.Category.XML, "GlobalPreferencesParser: $rpcResult")

                null
            }
        }
    }
}
